### PR TITLE
Fix validation errors and issues in shader objects sample

### LINKED ISSUE #843: Validation layer errors in shader object sample with recent SDKs
The recent LunarG Vulkan SDK includes an updated version of the validation layer that adds additional checks for the shader object extension. Running that sample with those (newer) layers results in a lot of validation errors that should be fixed.

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR fixes some validation layers in the shader objects sample that were introduced with recent Vulkan SDKs. It also fixes some general problems with the code that lead to resources not being properly freed (resulted in validation errors when closing the sample).

With these fixes and the latest SDK, the example does not longer show any validation errors, neither at startup nor at closing.

Fixes #843

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.md#General-Requirements)

### CODE REVIEW

File: samples/extensions/shader_object/shader_object.cpp [hunk 1]
```diff
@@ -171,10 +172,6 @@ bool ShaderObject::prepare(const vkb::ApplicationOptions &options)
 	camera.set_rotation({19.f, 312.f, 0.f});
 	camera.set_perspective(60.f, static_cast<float>(width) / static_cast<float>(height), 1024.f, 0.1f);
 
-	// For UI need to setup FBO and render pass
-	setup_framebuffer();
-	setup_render_pass();
```
Comment (tomadamatkinson) [inline samples/extensions/shader_object/shader_object.cpp:176]: Are these calls no longer needed?

Comment (gpx1000) [reply samples/extensions/shader_object/shader_object.cpp:176]: They are called elsewhere, I believe this just removes the second call.

File: samples/extensions/shader_object/shader_object.cpp [hunk 2]
```diff
@@ -423,6 +421,7 @@ void ShaderObject::load_assets()
 	VK_CHECK(vkCreateSampler(get_device().get_handle(), &sampler_create_info, nullptr, &heightmap_texture.sampler));
 
 	// Setup a repeating sampler for the terrain texture layers
+	vkDestroySampler(get_device().get_handle(), terrain_array_textures.sampler, nullptr);
```
Comment (gpx1000) [inline samples/extensions/shader_object/shader_object.cpp:424]: Is this now leaked?

File: samples/extensions/shader_object/shader_object.cpp (hunk 1 above)
Comment (SaschaWillems) [reply samples/extensions/shader_object/shader_object.cpp:176]: Exactly. The calls in this spot were not only unnecessary, but also resulted in resources getting acquired multiple times without getting properly freed.

File: samples/extensions/shader_object/shader_object.cpp (hunk 2 above)
Comment (SaschaWillems) [reply samples/extensions/shader_object/shader_object.cpp:424]: Can you elaborate? The sample did leak some stuff, and e.g. this change makes sure all resources are properly freed. I made sure that I didn't get any validation errors at instance destruction.

**Review by JoseEmilio-ARM — APPROVED** [review]

File: samples/extensions/shader_object/shader_object.cpp [hunk 3]
```diff
@@ -410,6 +407,7 @@ void ShaderObject::load_assets()
 	VkSamplerCreateInfo sampler_create_info = vkb::initializers::sampler_create_info();
 
 	// Setup a mirroring sampler for the height map
+	vkDestroySampler(get_device().get_handle(), heightmap_texture.sampler, nullptr);
```
Comment (asuessenbach) [inline samples/extensions/shader_object/shader_object.cpp:410]: That looks a bit obscure, destroying a sampler right before you create it.
Actually, that sampler has been created right before in `load_texture`... Maybe we should adjust the load_texture*-functions to optionally _not_ create a default sampler?

**Review by asuessenbach — COMMENTED** [review]
I still get some validation layer warnings on exit:
`Validation Warning: [ CHASSIS ] Object 0: handle = 0xe647ea0000000090, type = VK_OBJECT_TYPE_IMAGE; Object 1: handle = 0xd68cbb0000000091, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0x7366b7dd | VK Object VkImage 0xe647ea0000000090[] still has a reference to mem obj VkDeviceMemory 0xd68cbb0000000091[].
`
They'll be gone if you `vkDestroyImage` first and `vkFreeMemory` second in the ShaderObject destructor.

File: samples/extensions/shader_object/shader_object.cpp (hunk 3 above)
Comment (SaschaWillems) [reply samples/extensions/shader_object/shader_object.cpp:410]: Yeah, but I actually didn't want to change anything in the framework. I'll take a look and if it isn't too much work I'll try to fix it.

I'll take a look at the other message, which I oddly don't get with the current SDK :/

Comment (asuessenbach) [reply samples/extensions/shader_object/shader_object.cpp:410]: > but I actually didn't want to change anything in the framework.

Maybe just file an issue to improve that later on?

**Review by gpx1000 — APPROVED** [review]

**Comment by marty-johnson59** [comment]
Sascha will create an issue to address Andreas' last comment.  Merging based on discussion on Dec 18th meeting

File: samples/extensions/shader_object/shader_object.cpp (hunk 3 above)
Comment (SaschaWillems) [reply samples/extensions/shader_object/shader_object.cpp:410]: See #866 866

**Pull request merged**